connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime;
+
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.JmxReporter;
+import org.apache.kafka.common.metrics.MetricConfig;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.utils.AppInfoParser;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * The Connect metrics with JMX reporter.
+ */
+public class ConnectMetrics {
+
+    public static final String JMX_PREFIX = "kafka.connect";
+    public static final String WORKER_ID_TAG_NAME = "worker-id";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConnectMetrics.class);
+    private static final AtomicInteger CONNECT_WORKER_ID_SEQUENCE = new AtomicInteger(1);
+
+    private final Metrics metrics;
+    private final Time time;
+    private final String workerId;
+    private final ConcurrentMap<String, MetricGroup> groupsByName = new ConcurrentHashMap<>();
+
+    /**
+     * Create an instance.
+     *
+     * @param workerId the worker identifier; may be null
+     * @param config   the worker configuration; may not be null
+     * @param time     the time; may not be null
+     */
+    public ConnectMetrics(String workerId, WorkerConfig config, Time time) {
+        this.workerId = workerId != null ? makeValidName(workerId) : "worker-" + CONNECT_WORKER_ID_SEQUENCE.getAndIncrement();
```

Comment:
It's definitely null in tests where each test uses a different workerId in the metrics. This seemed to reflect other metric tests, so I went with it.

Reply:
hmm, i guess it seemed weird since that's not the intended usage and seems like a hack for the tests. not the end of the world, but seemed weird to me. which tests are you referring to so i can compare?